### PR TITLE
Chapter2

### PR DESCRIPTION
**Pull request by will-clarke**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -349,7 +364,8 @@ Implement a function that returns only the first half of a given list.
 "b"
 -}
 -- PUT THE FUNCTION TYPE IN HERE
-firstHalf l = error "firstHalf: Not implemented!"
+firstHalf :: [a] -> [a]
+firstHalf xs = take n xs where n = length xs `div` 2
```
Comment (chshersh) [inline src/Chapter2.hs:368]: Great usage of `where` construction! Makes code easily readable 💯

File: src/Chapter2.hs [hunk 2]
```diff
@@ -621,7 +640,10 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [Int] -> [Int]
+takeEven []           = []
+takeEven (x : _ : xs) = x : takeEven xs
+takeEven x            = x
```
Comment (chshersh) [inline src/Chapter2.hs:646]: I would recommend here to pattern match on the last remaining case explicitly to make use of GHC pattern matching checker :slightly_smiling_face: Alternatively, you can combine this case with the first line and reduce the number of lines in the implementation of this function
```suggestion
takeEven [x]          = [x]
```

File: src/Chapter2.hs [hunk 3]
```diff
@@ -728,7 +750,7 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate xs = concatMap (\ x -> replicate x x) xs
```
Comment (chshersh) [inline src/Chapter2.hs:753]: Now, after you've mastered eta-reduction, you can apply such a technique to this function as well  🙂 
But your implementation is already great 💯

File: src/Chapter2.hs [hunk 4]
```diff
@@ -842,7 +865,9 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [a] -> [a]
+rotate n xs | n < 0     = []
+            | otherwise = take (length xs) $ drop n $ cycle xs
```
Comment (chshersh) [inline src/Chapter2.hs:870]: That's a correct solution! Though, it does a little bit more than work when the given number is big and the list is small. You will cycle through the list a lot. Maybe you can think of a way to optimize this :slightly_smiling_face:

File: src/Chapter2.hs [hunk 5]
```diff
@@ -858,7 +883,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind []       = []
+rewind (x : xs) = rewind xs ++ [x]
```
Comment (chshersh) [inline src/Chapter2.hs:888]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by chshersh — APPROVED** [review]
Nice work on Chapter Two tasks!
And well done with the advanced task :+1: 

I left some comments with suggestions and general feedback :slightly_smiling_face:

**Comment by vrom911** [comment]
Thanks for participating in Learn4Haskell! Hope you enjoyed this Hacktoberfest 🥳 

I am going to close this PR as we can't merge the solutions into our course 🔐 

Even though the Hacktoberfest is over, we are going to continue reviewing, so feel free to open new PRs, if you haven't finished the course yet! 🎊

**Pull request closed without merging**